File: plugins/obsidian-memory/scripts/statusline.py
#!/usr/bin/env python3
"""Claude Code statusline: shows obsidian-memory plugin's token share for the
current session. Mirrors the accounting in the plugin's /usage script.

Stdin: Claude Code session JSON (session_id, transcript_path, cwd, ...).
Stdout: one line, e.g. "obsidian-memory 18.2k tok · 4.3%"
       When cwd's repo is registered+enabled in repos.json, the prefix
       becomes "obsidian-memory • <project>" so the project tag is visible
       at a glance.
"""
from __future__ import annotations
import json, os, subprocess, sys
from datetime import datetime, timezone
from pathlib import Path

USAGE_DIR = os.environ.get("MEMORY_USAGE_DIR", "/tmp/claude-memory-usage")
CHARS_KINDS = ("session_start", "gate_inject")


def _project_tag(cwd: str) -> str:
    """Return the registered project name for cwd's repo, or '' if not
    registered+enabled. Best-effort: any error → ''."""
    if not cwd:
        return ""
    try:
        toplevel = subprocess.run(
            ["git", "-C", cwd, "rev-parse", "--show-toplevel"],
            capture_output=True, text=True, timeout=2,
        )
    except (FileNotFoundError, subprocess.SubprocessError):
        return ""
    if toplevel.returncode != 0:
        return ""
    repo_root = toplevel.stdout.strip()
    if not repo_root:
        return ""
    repos_file = os.environ.get(
        "OBSIDIAN_MEMORY_REPOS_FILE",
        str(Path.home() / ".config/obsidian-memory/repos.json"),
    )
    try:
        with open(repos_file, encoding="utf-8") as f:
            data = json.load(f)
    except (FileNotFoundError, json.JSONDecodeError, OSError):
        return ""
    entry = (data.get("repos") or {}).get(str(Path(repo_root).resolve()))
    if not entry or not entry.get("enabled"):
        return ""
    return str(entry.get("project") or "")


def parse_ts(s: str):
    if not s:
        return None
    if s.endswith("Z"):
        s = s[:-1] + "+00:00"
    try:
        dt = datetime.fromisoformat(s)
    except ValueError:
        return None
    if dt.tzinfo is None:
        dt = dt.astimezone(timezone.utc)
    return dt


def fmt_tok(n: int) -> str:
    if n >= 1_000_000:
        return f"{n/1_000_000:.1f}M"
    if n >= 1_000:
        return f"{n/1_000:.1f}k"
    return str(n)


def render(text: str = "", project: str = "") -> None:
    # Obsidian-purple (#A78BFA) prefix via 24-bit truecolor.
    prefix = "\033[38;2;167;139;250mobsidian-memory\033[0m"
    if project:
        # Dim bullet + project name in default color.
        prefix = f"{prefix} \033[2m•\033[0m {project}"
    if text:
        sys.stdout.write(f"{prefix} {text}")
    else:
        sys.stdout.write(f"{prefix} \033[2m—\033[0m")
    sys.stdout.write("\n")


def main() -> None:
    try:
        payload = json.load(sys.stdin)
    except Exception:
        render()
        return

    session_id = payload.get("session_id") or ""
    transcript = payload.get("transcript_path") or ""
    cwd = payload.get("cwd") or os.getcwd()
    project = _project_tag(cwd)

    if not session_id:
        render(project=project)
        return

    safe_id = "".join(c if c.isalnum() or c in "._-" else "_" for c in session_id)
    plugin_log = os.path.join(USAGE_DIR, f"{safe_id}.jsonl")

    # Plugin-side events
    api_sum = 0
    chars_events = []  # (approx_tokens, ts)
    if os.path.isfile(plugin_log):
        try:
            with open(plugin_log) as f:
                for line in f:
                    line = line.strip()
                    if not line:
                        continue
                    try:
                        e = json.loads(line)
                    except Exception:
                        continue
                    kind = e.get("kind", "")
                    if e.get("mode") == "api":
                        u = e.get("usage", {}) or {}
                        api_sum += (
                            int(u.get("input_tokens", 0) or 0)
                            + int(u.get("output_tokens", 0) or 0)
                            + int(u.get("cache_read_input_tokens", 0) or 0)
                            + int(u.get("cache_creation_input_tokens", 0) or 0)
                        )
                    elif kind in CHARS_KINDS:
                        tok = int(e.get("approx_tokens", 0) or 0)
                        ts = parse_ts(e.get("ts", ""))
                        chars_events.append((tok, ts))
        except Exception:
            pass

    # Main transcript usage (deduped by message.id to avoid snapshot/replay double-count)
    main_sum = 0
    main_msgs_ts = []
    if transcript and os.path.isfile(transcript):
        seen = set()
        try:
            with open(transcript) as f:
                for line in f:
                    try:
                        e = json.loads(line)
                    except Exception:
                        continue
                    msg = e.get("message") or {}
                    usage = msg.get("usage")
                    if not usage:
                        continue
                    mid = msg.get("id")
                    if mid:
                        if mid in seen:
                            continue
                        seen.add(mid)
                    t = parse_ts(e.get("timestamp", ""))
                    if t is not None:
                        main_msgs_ts.append(t)
                    main_sum += (
                        int(usage.get("input_tokens", 0) or 0)
                        + int(usage.get("output_tokens", 0) or 0)
                        + int(usage.get("cache_read_input_tokens", 0) or 0)
                        + int(usage.get("cache_creation_input_tokens", 0) or 0)
                    )
        except Exception:
            pass

    # Injection × turns-alive attribution
    n_msgs = len(main_msgs_ts)
    plugin_main_attr = 0
    for tok, ts in chars_events:
        if ts is None:
            turns_alive = n_msgs
        else:
            turns_alive = sum(1 for mt in main_msgs_ts if mt >= ts)
        plugin_main_attr += tok * turns_alive

    plugin_total = plugin_main_attr + api_sum
    total_session = main_sum + api_sum

    if total_session <= 0 or plugin_total <= 0:
        render(project=project)
        return

    share = plugin_total / total_session * 100
    # Color the share dim/yellow/red by magnitude — passive cost signal.
    if share >= 25:
        color = "\033[31m"  # red
    elif share >= 10:
        color = "\033[33m"  # yellow
    else:
        color = "\033[2m"   # dim
    reset = "\033[0m"
    render(f"{fmt_tok(plugin_total)} tok · {color}{share:.1f}%{reset}", project=project)


if __name__ == "__main__":
    main()

File: src/statusline.rs
//! Claude Code statusline — port of scripts/statusline.py.
//!
//! Reads Claude Code session JSON on stdin, scans the obsidian-memory plugin
//! usage log + main transcript, attributes plugin token cost via
//! injection×turns-alive, and renders one colored line.
//!
//! Self-disables when:
//!   - the plugin is uninstalled (no `enabledPlugins.obsidian-memory@*: true`)
//!   - `OBSIDIAN_MEMORY_STATUSLINE_ENABLED=false` (via env or config.env)
//!
//! Both produce *zero* output, matching Python's `return` path.

use std::io::{Read, Write};
use std::path::{Path, PathBuf};
use std::process::Command;

use anyhow::Result;
use chrono::{DateTime, Utc};
use serde_json::Value;

const CHARS_KINDS: &[&str] = &["session_start", "gate_inject"];

pub fn run() -> Result<i32> {
    if !plugin_installed() {
        return Ok(0);
    }
    if !config_flag("OBSIDIAN_MEMORY_STATUSLINE_ENABLED", true) {
        return Ok(0);
    }

    let mut payload_text = String::new();
    let _ = std::io::stdin().read_to_string(&mut payload_text);
    let payload: Value = match serde_json::from_str(&payload_text) {
        Ok(v) => v,
        Err(_) => {
            render(None, None);
            return Ok(0);
        }
    };

    let session_id = payload.get("session_id").and_then(|v| v.as_str()).unwrap_or("").to_string();
    let transcript = payload.get("transcript_path").and_then(|v| v.as_str()).unwrap_or("").to_string();
    let cwd = payload
        .get("cwd")
        .and_then(|v| v.as_str())
        .map(|s| s.to_string())
        .unwrap_or_else(|| std::env::current_dir().map(|p| p.to_string_lossy().into_owned()).unwrap_or_default());

    let project = project_tag(&cwd);

    if session_id.is_empty() {
        render(None, project.as_deref());
        return Ok(0);
    }

    let safe_id: String = session_id
        .chars()
        .map(|c| if c.is_ascii_alphanumeric() || matches!(c, '.' | '_' | '-') { c } else { '_' })
        .collect();
    let usage_dir = std::env::var("MEMORY_USAGE_DIR").unwrap_or_else(|_| "/tmp/claude-memory-usage".into());
    let plugin_log = PathBuf::from(usage_dir).join(format!("{safe_id}.jsonl"));

    let (api_sum, chars_events) = read_plugin_log(&plugin_log);
    let (main_sum, main_msgs_ts) = read_main_transcript(&transcript);

    // Attribution: each chars-event injected `tok` tokens into `turns_alive` turns.
    let n_msgs = main_msgs_ts.len();
    let mut plugin_main_attr: u64 = 0;
    for (tok, ts) in &chars_events {
        let turns_alive = match ts {
            Some(ts) => main_msgs_ts.iter().filter(|mt| *mt >= ts).count(),
            None => n_msgs,
        };
        plugin_main_attr += (*tok) * (turns_alive as u64);
    }

    let plugin_total = plugin_main_attr + api_sum;
    let total_session = main_sum + api_sum;

    if total_session == 0 || plugin_total == 0 {
        render(None, project.as_deref());
        return Ok(0);
    }

    let share = (plugin_total as f64) / (total_session as f64) * 100.0;
    // Color by magnitude — passive cost signal.
    let color = if share >= 25.0 {
        "\x1b[31m" // red
    } else if share >= 10.0 {
        "\x1b[33m" // yellow
    } else {
        "\x1b[2m" // dim
    };
    let body = format!(
        "{} tok · {color}{share:.1}%\x1b[0m",
        fmt_tok(plugin_total),
    );
    render(Some(&body), project.as_deref());
    Ok(0)
}

/// True iff `~/.claude/settings.json`'s `enabledPlugins` has any
/// `obsidian-memory@<ver>` key with truthy value.
fn plugin_installed() -> bool {
    let Some(home) = dirs::home_dir() else { return false };
    let settings_file = home.join(".claude/settings.json");
    if !settings_file.is_file() {
        return false;
    }
    let Ok(text) = std::fs::read_to_string(&settings_file) else { return false };
    let Ok(data): std::result::Result<Value, _> = serde_json::from_str(&text) else { return false };
    let Some(enabled) = data.get("enabledPlugins").and_then(|v| v.as_object()) else { return false };
    enabled
        .iter()
        .any(|(k, v)| k.starts_with("obsidian-memory@") && truthy(v))
}

fn truthy(v: &Value) -> bool {
    match v {
        Value::Bool(b) => *b,
        Value::Null => false,
        Value::Number(n) => n.as_f64().map(|x| x != 0.0).unwrap_or(false),
        Value::String(s) => !s.is_empty(),
        Value::Array(a) => !a.is_empty(),
        Value::Object(o) => !o.is_empty(),
    }
}

/// Read a boolean flag from config.env. Honors env override first.
/// Parses minimal `KEY=value` and `export KEY="value"` lines.
fn config_flag(name: &str, default: bool) -> bool {
    if let Ok(env_val) = std::env::var(name) {
        return matches!(env_val.trim().to_lowercase().as_str(), "true" | "1" | "yes" | "on");
    }
    let config_path = match std::env::var("OBSIDIAN_MEMORY_CONFIG_FILE") {
        Ok(p) => PathBuf::from(p),
        Err(_) => match dirs::home_dir() {
            Some(h) => h.join(".config/obsidian-memory/config.env"),
            None => return default,
        },
    };
    if !config_path.is_file() {
        return default;
    }
    let Ok(text) = std::fs::read_to_string(&config_path) else { return default };
    for raw in text.lines() {
        let mut line = raw.trim();
        if line.is_empty() || line.starts_with('#') {
            continue;
        }
        if let Some(rest) = line.strip_prefix("export ") {
            line = rest.trim_start();
        }
        let Some(eq) = line.find('=') else { continue };
        let (k, v) = line.split_at(eq);
        if k.trim() != name {
            continue;
        }
        let val = v[1..].trim();
        // Strip matching quote pair if present.
        let unquoted = if val.len() >= 2 {
            let first = val.as_bytes()[0] as char;
            let last = val.as_bytes()[val.len() - 1] as char;
            if (first == '\'' || first == '"') && first == last {
                &val[1..val.len() - 1]
            } else {
                val
            }
        } else {
            val
        };
        return matches!(unquoted.to_lowercase().as_str(), "true" | "1" | "yes" | "on");
    }
    default
}

/// Resolve a registered+enabled project name for `cwd`'s git repo, or `None`.
fn project_tag(cwd: &str) -> Option<String> {
    if cwd.is_empty() {
        return None;
    }
    let mut cmd = Command::new("git");
    cmd.arg("-C").arg(cwd).args(["rev-parse", "--show-toplevel"]);
    let Ok(out) = cmd.output() else { return None };
    if !out.status.success() {
        return None;
    }
    let project_root = String::from_utf8_lossy(&out.stdout).trim().to_string();
    if project_root.is_empty() {
        return None;
    }
    let projects_file = match std::env::var("OBSIDIAN_MEMORY_PROJECTS_FILE") {
        Ok(p) => PathBuf::from(p),
        Err(_) => match dirs::home_dir() {
            Some(h) => h.join(".config/obsidian-memory/projects.json"),
            None => return None,
        },
    };
    let Ok(text) = std::fs::read_to_string(&projects_file) else { return None };
    let Ok(data): std::result::Result<Value, _> = serde_json::from_str(&text) else { return None };
    // Python: str(Path(project_root).resolve()) — symlink-resolved absolute.
    let resolved_key = crate::vault::walk::absolute(Path::new(&project_root))
        .to_string_lossy()
        .into_owned();
    let entry = data
        .get("projects")
        .and_then(|v| v.as_object())
        .and_then(|m| m.get(&resolved_key))?;
    if !truthy(entry.get("enabled").unwrap_or(&Value::Bool(false))) {
        return None;
    }
    let project = entry.get("project").and_then(|v| v.as_str()).unwrap_or("");
    if project.is_empty() {
        None
    } else {
        Some(project.to_string())
    }
}

type CharsEvent = (u64, Option<DateTime<Utc>>);

fn read_plugin_log(path: &Path) -> (u64, Vec<CharsEvent>) {
    let mut api_sum: u64 = 0;
    let mut chars_events: Vec<CharsEvent> = Vec::new();
    let Ok(text) = std::fs::read_to_string(path) else { return (api_sum, chars_events); };
    for raw in text.lines() {
        let line = raw.trim();
        if line.is_empty() {
            continue;
        }
        let Ok(e) = serde_json::from_str::<Value>(line) else { continue };
        let kind = e.get("kind").and_then(|v| v.as_str()).unwrap_or("");
        if e.get("mode").and_then(|v| v.as_str()) == Some("api") {
            let u = e.get("usage").cloned().unwrap_or(Value::Null);
            api_sum += token_total(&u);
        } else if CHARS_KINDS.contains(&kind) {
            let tok = e.get("approx_tokens").and_then(|v| v.as_u64()).unwrap_or(0);
            let ts = e.get("ts").and_then(|v| v.as_str()).and_then(parse_ts);
            chars_events.push((tok, ts));
        }
    }
    (api_sum, chars_events)
}

fn read_main_transcript(path: &str) -> (u64, Vec<DateTime<Utc>>) {
    let mut main_sum: u64 = 0;
    let mut msgs_ts: Vec<DateTime<Utc>> = Vec::new();
    if path.is_empty() { return (main_sum, msgs_ts); }
    let p = Path::new(path);
    if !p.is_file() { return (main_sum, msgs_ts); }
    let Ok(text) = std::fs::read_to_string(p) else { return (main_sum, msgs_ts); };
    let mut seen = std::collections::HashSet::<String>::new();
    for raw in text.lines() {
        let Ok(e) = serde_json::from_str::<Value>(raw) else { continue };
        let msg = e.get("message").cloned().unwrap_or(Value::Null);
        let usage = msg.get("usage").cloned().unwrap_or(Value::Null);
        if usage.is_null() { continue; }
        if let Some(mid) = msg.get("id").and_then(|v| v.as_str()) {
            if !seen.insert(mid.to_string()) {
                continue;
            }
        }
        if let Some(ts) = e.get("timestamp").and_then(|v| v.as_str()).and_then(parse_ts) {
            msgs_ts.push(ts);
        }
        main_sum += token_total(&usage);
    }
    (main_sum, msgs_ts)
}

fn token_total(usage: &Value) -> u64 {
    let g = |k: &str| usage.get(k).and_then(|v| v.as_u64()).unwrap_or(0);
    g("input_tokens") + g("output_tokens") + g("cache_read_input_tokens") + g("cache_creation_input_tokens")
}

fn parse_ts(s: &str) -> Option<DateTime<Utc>> {
    if s.is_empty() { return None; }
    // Python: replace trailing "Z" with "+00:00", then fromisoformat.
    let normalized = if let Some(stripped) = s.strip_suffix('Z') {
        format!("{stripped}+00:00")
    } else {
        s.to_string()
    };
    DateTime::parse_from_rfc3339(&normalized).ok().map(|dt| dt.with_timezone(&Utc))
}

fn fmt_tok(n: u64) -> String {
    if n >= 1_000_000 {
        format!("{:.1}M", n as f64 / 1_000_000.0)
    } else if n >= 1_000 {
        format!("{:.1}k", n as f64 / 1_000.0)
    } else {
        n.to_string()
    }
}

fn render(text: Option<&str>, project: Option<&str>) {
    // Truecolor obsidian-purple #A78BFA prefix.
    let mut prefix = String::from("\x1b[38;2;167;139;250mobsidian-memory\x1b[0m");
    if let Some(p) = project {
        if !p.is_empty() {
            prefix.push_str(&format!(" \x1b[2m•\x1b[0m {p}"));
        }
    }
    let stdout = std::io::stdout();
    let mut h = stdout.lock();
    let body = match text {
        Some(t) if !t.is_empty() => format!("{prefix} {t}\n"),
        _ => format!("{prefix} \x1b[2m—\x1b[0m\n"),
    };
    let _ = h.write_all(body.as_bytes());
}
